src/cqueue.rs

```diff
@@ -125,44 +152,42 @@ impl CompletionQueue<'_> {
 
     #[cfg(feature = "unstable")]
     #[inline]
-    pub fn fill<'a>(&mut self, entries: &'a mut [MaybeUninit<Entry>]) -> &'a mut [Entry] {
+    pub fn fill<'a>(&mut self, entries: &'a mut [MaybeUninit<E>]) -> &'a mut [E] {
         let len = std::cmp::min(self.len(), entries.len());
 
         for entry in &mut entries[..len] {
-            *entry = MaybeUninit::new(Entry(unsafe {
-                *self
-                    .queue
-                    .cqes
-                    .add((self.head & self.queue.ring_mask) as usize)
-            }));
-            self.head = self.head.wrapping_add(1);
+            entry.write(self.pop());
         }
 
-        unsafe { std::slice::from_raw_parts_mut(entries as *mut _ as *mut Entry, len) }
+        unsafe { std::slice::from_raw_parts_mut(entries as *mut _ as *mut E, len) }
+    }
+
+    fn pop(&mut self) -> E {
```

Comment:
This should be marked `unsafe` and `inline`.

Reply:
I don't think this needs `unsafe`: it is just reading some bytes from a region of memory we know is valid. There is no additional contract the caller must follow to avoid undefined behavior.

Should I make it `unsafe` nonetheless?

Reply:
I'm not sure it's always a valid memory region, if kernel doesn't write an entry into it, it's probably uninitialized memory.